### DISCUSSION TITLE
make Field.defaultOperator type more strict

### DISCUSSION THREAD

**Discussion by hekystyle** [react-querybuilder/react-querybuilder#553] [Ideas]
Hello, in file `packages/react-querybuilder/src/types/basic.ts` there is type called `Field` with property `defaultOperator` and I feel like type should be generic param `OperatorName` instead of `string`, do you think this change is valid?

Example here https://github.com/react-querybuilder/react-querybuilder/compare/main...hekystyle:react-querybuilder:improve-types, please ignore formatting changes, I will revert them before making PR.

**Comment by jakeboone02** [comment]
Thanks for bringing this up. I'm actually in the middle of an attempt to improve type inference for RQB, so this will help. I'll include it if you make a PR.

I think we should probably type the `operator` parameter in the `valueEditorType` and `valueSources` functions as well:

```diff
-  valueEditorType?: ValueEditorType | ((operator: string) => ValueEditorType);
-  valueSources?: ValueSources | ((operator: string) => ValueSources);
+  valueEditorType?: ValueEditorType | ((operator: OperatorName) => ValueEditorType);
+  valueSources?: ValueSources | ((operator: OperatorName) => ValueSources);
```

That will guarantee that the `operator` argument will be limited to the `OperatorName` type, even if the user's provided function accepts `string`. Feel free to include that change in your PR.
